diff --git a/proto/meraki/meraki.c b/proto/meraki/meraki.c
index 5a508a23..352518b1 100644
--- a/proto/meraki/meraki.c
+++ b/proto/meraki/meraki.c
@@ -20,10 +20,10 @@
 #include "proto/bgp/bgp.h"
 
 #include "meraki.h"
+#include "meraki_bird_defines.hh"
 
 #include <stdio.h>
 
-#define BIRD_EXPORTED_ROUTE_BEGIN 100000 //see meraki/wired_brain.cc
 #define ASPATH_BUF_LEN 512
 //Accommodate worst-case nexthop size
 #define NEXTHOP_BUF_LEN sizeof("[\"xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx:xxxx\", "\
@@ -120,8 +120,8 @@ meraki_notify(struct proto *P, rtable *src_table, net *n, rte *new, rte *old, ea
    unsigned peer_num, route_count = BIRD_EXPORTED_ROUTE_BEGIN;
    char ip_buf[STD_ADDRESS_P_LENGTH], source_key[BGP_SOURCE_KEY_LEN];
 
-   const char* outfile_tmp = "/tmp/config.tmp.tmp";
-   const char* outfile = "/tmp/config.tmp";
+   const char* outfile_tmp = BIRD_ROUTES_EXPORT_FILE ".tmp";
+   const char* outfile = BIRD_ROUTES_EXPORT_FILE;
 
    FILE *out = fopen(outfile_tmp, "w");
    if (!out) {
@@ -129,8 +129,8 @@ meraki_notify(struct proto *P, rtable *src_table, net *n, rte *new, rte *old, ea
       return;
    }
 
-   const char* info_outfile_tmp = "/tmp/bird_routes.json.tmp";
-   const char* info_outfile = "/tmp/bird_routes.json";
+   const char* info_outfile_tmp = BIRD_ROUTES_JSON_FILE ".tmp";
+   const char* info_outfile = BIRD_ROUTES_JSON_FILE;
    int is_bgp_route = 0;
 
    FILE *info_out = fopen(info_outfile_tmp, "w");
-- 
2.22.0

