diff --git a/pam_ldap.h b/pam_ldap.h
index 554a63560f0..4b2d612aab8 100644
--- a/pam_ldap.h
+++ b/pam_ldap.h
@@ -134,6 +134,8 @@ typedef struct pam_ldap_config
     char *tls_key;
     /* tls randfile */
     char *tls_randfile;
+    /* enable/disable fips mode */
+    int tls_fips_mode;
     /* directory for debug files */
     char *logdir;
     /* ldap debug level */
diff --git a/pam_ldap.c b/pam_ldap.c
index 5920d0bf932..e673550b220 100644
--- a/pam_ldap.c
+++ b/pam_ldap.c
@@ -662,6 +662,7 @@ _alloc_config (pam_ldap_config_t ** presult)
   result->tls_cert = NULL;
   result->tls_key = NULL;
   result->tls_randfile = NULL;
+  result->tls_fips_mode = 0;
   result->logdir = NULL;
   result->sasl_mechanism = NULL;
   result->debug = 0;
@@ -1087,6 +1088,19 @@ _read_config (const char *configFile, pam_ldap_config_t ** presult)
 	{
 	  CHECKPOINTER (result->tls_randfile = strdup (v));
 	}
+      else if (!strcasecmp (k, "tls_fips_mode"))
+        {
+          if (!strcasecmp (v, "on") || !strcasecmp (v, "yes")
+              || !strcasecmp (v, "true"))
+            {
+              result->tls_fips_mode = 1;
+            }
+          else if (!strcasecmp (v, "off") || !strcasecmp (v, "no")
+                   || !strcasecmp (v, "false"))
+            {
+              result->tls_fips_mode = 0;
+            }
+        }
       else if (!strcasecmp (k, "logdir"))
 	{
 	  CHECKPOINTER (result->logdir = strdup (v));
@@ -1496,6 +1510,19 @@ _set_ssl_default_options (pam_ldap_session_t * session)
 	}
     }

+  if (session->conf->tls_fips_mode > -1)
+    {
+      rc = ldap_set_option (NULL, LDAP_OPT_X_TLS_FIPS_MODE,
+                            &session->conf->tls_fips_mode);
+      if (rc != LDAP_SUCCESS)
+        {
+          syslog (LOG_ERR,
+                  "pam_ldap: ldap_set_option(LDAP_OPT_X_TLS_FIPS_MODE): %s",
+                  ldap_err2string (rc));
+          return LDAP_OPERATIONS_ERROR;
+        }
+    }
+
   return LDAP_SUCCESS;
 }

