From 505b95ae25bd2bdbba15f516d5242c03cdb333a8 Mon Sep 17 00:00:00 2001
From: Felix Fietkau <nbd@openwrt.org>
Date: Mon, 14 May 2007 16:05:17 +0000
Subject: [PATCH]     make udhcpc silent after 3 'Sending discover ...'
 messages

    SVN-Revision: 7227
---
 networking/udhcp/dhcpc.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/networking/udhcp/dhcpc.c b/networking/udhcp/dhcpc.c
index 6b8dfae..676b117 100644
--- a/networking/udhcp/dhcpc.c
+++ b/networking/udhcp/dhcpc.c
@@ -727,6 +727,7 @@ static int bcast_or_ucast(struct dhcp_packet *packet, uint32_t ciaddr, uint32_t
 static NOINLINE int send_discover(uint32_t requested)
 {
 	struct dhcp_packet packet;
+	static int msgs = 0;
 
 	/* Fill in: op, htype, hlen, cookie, chaddr fields,
 	 * xid field, message type option:
@@ -741,6 +742,7 @@ static NOINLINE int send_discover(uint32_t requested)
 	 */
 	add_client_options(&packet);
 
+	if (msgs++ < 3)
 	bb_simple_info_msg("broadcasting discover");
 	return raw_bcast_from_client_data_ifindex(&packet, INADDR_ANY);
 }
-- 
2.32.0

